File: (deprecated)MediaScraper/MediaScraperHtml/ThreeMbLinksScraper.cs
using HtmlAgilityPack;
using MediaHatchCore;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace MediaScraperHtml
{
	public class ThreeMbLinksScraper
	{
		public async Task ScrapeAsync(Media media, string mediaPageUrl)
		{
			HtmlNode contentNode = ThreeMbMediaScraper.GetContentNode(await ThreeMbMediaScraper.GetHtmlDocumentAsync(mediaPageUrl));
			if (contentNode != null)
			{
				HtmlNode entryNode = contentNode.ChildNodes.First(c => c.Name == "div")?.ChildNodes.First(c => c.Name == "div" && c.GetAttributeValue("class", string.Empty) == "entry");
				if (entryNode != null)
				{
					HtmlNode linksNode = entryNode.ChildNodes.First(c => c.Name == "p" && CultureInfo.InvariantCulture.CompareInfo.IndexOf(c.InnerText, "click here to get links", CompareOptions.IgnoreCase) >= 0);
					if (linksNode != null)
					{
						string linksPageUrl = linksNode.ChildNodes.First(c => c.Name == "a")?.GetAttributeValue("href", null);
						if (!string.IsNullOrEmpty(linksPageUrl))
						{
							HtmlDocument linksPageDocument = await GetLinksPageHtmlAsync(linksPageUrl);
							HtmlNode linksTableNode = linksPageDocument.DocumentNode.ChildNodes.First(c => c.Name == "html")?.ChildNodes.First(c => c.Name == "body")?.ChildNodes
														.First(c => c.Name == "div" && c.Id == "page")?.ChildNodes.First(c => c.Name == "center")?.ChildNodes
														.First(c => c.Name == "div")?.ChildNodes.First(c => c.Name == "div")?.ChildNodes.First(c => c.Name == "p")?.ChildNodes
														.First(c => c.Name == "center")?.ChildNodes.First(c => c.Name == "table")?.ChildNodes.First(c => c.Name == "tbody");
							List<HtmlNode> linksPackagesNodes = new List<HtmlNode>(linksTableNode.ChildNodes.Where(c => c.Name == "tr"));

							// remove first and last links packages nodes (Direct download & Single links)
							linksPackagesNodes.RemoveAt(0);
							linksPackagesNodes.RemoveAt(linksPackagesNodes.Count - 1);

							media.LinksPackages = RetrieveLinksPackages(linksPackagesNodes);
						}
					}
				}
			}
		}

		private async Task<HtmlDocument> GetLinksPageHtmlAsync(string linksPageUrl)
		{
			ASCIIEncoding encoding = new ASCIIEncoding();
			string postData = "post_password=300mbfilms&Submit=Submit";
			byte[] data = encoding.GetBytes(postData);

			HttpWebRequest request = (HttpWebRequest)WebRequest.Create("http://earn-money-onlines.info/wp-login.php?action=postpass");
			request.CookieContainer = new CookieContainer();
			request.Method = "POST";
			request.ContentType = "application/x-www-form-urlencoded";
			using (Stream newStream = await request.GetRequestStreamAsync())
				newStream.Write(data, 0, data.Length);

			WebResponse response = await request.GetResponseAsync();
			response.Dispose();

			CookieCollection cookies = request.CookieContainer.GetCookies(new Uri("http://earn-money-onlines.info"));

			Cookie authenticationCookie = null;
			int cookieIndex = 0;
			foreach (Cookie cookie in cookies)
			{
				if (cookieIndex == 1)
				{
					authenticationCookie = cookie;
					break;
				}
				cookieIndex++;
			}

            HttpWebRequest linksRequest = HttpWebRequest.CreateHttp(linksPageUrl);
            linksRequest.CookieContainer = request.CookieContainer;
            response = await linksRequest.GetResponseAsync();
            String responseString = string.Empty;
            using (Stream s = response.GetResponseStream())
            {
                StreamReader reader = new StreamReader(s, Encoding.UTF8);
                responseString = reader.ReadToEnd();
            }
            return ThreeMbMediaScraper.GetHtmlDocument(responseString);
		}

		private List<LinksPackage> RetrieveLinksPackages(List<HtmlNode> linksPackagesNodes)
		{
			List<LinksPackage> linksPackages = new List<LinksPackage>();
			foreach (HtmlNode node in linksPackagesNodes)
			{
				LinksPackage newLinksPackage = RetrieveLinksPackage(node);
				if (newLinksPackage != null)
					linksPackages.Add(newLinksPackage);
			}
			return linksPackages;
		}

		private LinksPackage RetrieveLinksPackage(HtmlNode linksPackageNode)
		{
			List<HtmlNode> linksNodes = new List<HtmlNode>(linksPackageNode.ChildNodes.Where(c => c.Name == "td"));
			// Remove description node
			linksNodes.RemoveAt(0);

			List<string> links = new List<string>();
			string fileHost = null;
			foreach (HtmlNode linkNode in linksNodes)
			{
				string newLink = RetrieveLink(linkNode, out fileHost);
				if (!string.IsNullOrEmpty(newLink))
					links.Add(newLink);
			}
			if (links.Count > 0 && !string.IsNullOrEmpty(fileHost))
				return new LinksPackage(fileHost, links);
			else
				return null;
		}

		private string RetrieveLink(HtmlNode linkNode, out string fileHost)
		{
			fileHost = null;
			string originalLink = linkNode.ChildNodes.First(c => c.Name == "a")?.GetAttributeValue("href", null);
			if (originalLink != null)
			{
				Match linkMatch = _linkRegex.Match(originalLink);
				if (linkMatch.Success)
				{
					string link = linkMatch.Groups[_linkGroupName].Value;
					fileHost = linkMatch.Groups[_fileHostGroupName].Value;
					if (!string.IsNullOrEmpty(link))
						return link;
				}
			}
			return originalLink;
		}

		private const string _linkGroupName = "link";
		private const string _fileHostGroupName = "filehost";

		private static Regex _linkRegex = new Regex($"(http:\\/\\/linkshrink.net.*)*(?<{_linkGroupName}>(http)(s)?:\\/\\/(?<{_fileHostGroupName}>[^\\/]*)\\/.*)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
	}
}

File: Backend/media-hatch-server/src/html/three-mb-links-scraper.ts
import { ThreeMbScrapedMediaInfo } from './three-mb-scraped-media-info';
import { ThreeMbMediaScraper } from './three-mb-media-scraper';

import { LinksPackage } from 'media-hatch-core';

import * as WebRequest from 'web-request';
import * as cheerio from 'cheerio';
import * as xregexp from 'xregexp';

export class ThreeMbLinksScraper {
  public scrape(scrapedMediaInfo: ThreeMbScrapedMediaInfo): Promise<LinksPackage[]> {
    return ThreeMbMediaScraper.getHtmlDocument(scrapedMediaInfo.pageUrl).then(htmlDocument => {
      const entry = htmlDocument.find('.entry');
      let linksPageNodeFound = false;
      const linksPageNode = entry.find('a').filter((i, el) => {
        return (
          cheerio
            .load(el)
            .root()
            .text()
            .toLowerCase()
            .indexOf('click here to get links') >= 0
        );
      });
      const linksPageUrl = linksPageNode.first().attr('href');
      return this.getLinksPageHtml(linksPageUrl).then(htmlDocument => {
        let unfilteredNodes = htmlDocument.find('tr');
        unfilteredNodes.first().remove();
        unfilteredNodes.last().remove();
        scrapedMediaInfo.media.linksPackages = this.retrieveLinksPackages(htmlDocument.find('tr'));
        return scrapedMediaInfo.media.linksPackages;
      });
    });
  }

  private getLinksPageHtml(linksPageUrl: string): Promise<Cheerio> {
    return WebRequest.post('http://earn-money-onlines.info/wp-login.php?action=postpass', { jar: true, form: 'post_password=300mbfilms&Submit=Submit' }).then(
      response => {
        return ThreeMbMediaScraper.getHtmlDocument(linksPageUrl, {
          jar: response.request.options.jar
        });
      }
    );
  }

  private retrieveLinksPackages(linksPackagesNodes: Cheerio): LinksPackage[] {
    let linksPackages = new Array<LinksPackage>();
    for (let i = 0; i < linksPackagesNodes.length; i++) {
      let newLinksPackage = this.retrieveLinksPackage(linksPackagesNodes.eq(i));
      if (newLinksPackage != null) {
        linksPackages.push(newLinksPackage);
      }
    }
    return linksPackages;
  }

  private retrieveLinksPackage(linksPackageNode: Cheerio): LinksPackage {
    const linksNodes = linksPackageNode.find('td');
    let links = new Array<string>();
    let fileHost: string;

    for (let j = 1; j < linksNodes.length; j++) {
      let newLink = this.retrieveLink(linksNodes.eq(j));
      if (newLink) {
        links.push(newLink);
        if (!fileHost) {
          const fileHostMatchArray = xregexp.exec(newLink, this.fileHostRegex);
          if (fileHostMatchArray) {
            fileHost = fileHostMatchArray['fileHost'];
          }
        }
      }
    }
    if (links.length > 0 && fileHost) {
      return new LinksPackage(fileHost, links);
    } else {
      return null;
    }
  }

  private retrieveLink(linkNode: Cheerio): string {
    const originalLink = linkNode.children('a').attr('href');
    if (originalLink) {
      const linkMatchArray = xregexp.exec(originalLink, this.linkRegex);
      if (linkMatchArray) {
        return linkMatchArray['link'];
      }
    }
    return undefined;
  }

  private linkRegex = xregexp.build('(http://linkshrink.net.*)*(?<link>{{0}})', ['(http)(s)?://.*'], 'i');

  private fileHostRegex = xregexp.build('(http)(s)?://(?<fileHost>{{0}})/.*', ['[^/]*'], 'i');
}
